pycbc/inference/sampler/ultranest.py

```diff
@@ -0,0 +1,184 @@
+# with this program; if not, write to the Free Software Foundation, Inc.,
+# 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+
+
+#
+# =============================================================================
+#
+#                                   Preamble
+#
+# =============================================================================
+#
+"""
+This modules provides classes and functions for using the dynesty sampler
+packages for parameter estimation.
+"""
+
+
+from __future__ import absolute_import
+
+import logging
+
+from pycbc.inference.io.ultranest import UltranestFile
+from .base import (BaseSampler, setup_output)
+from .base_cube import setup_calls
+
+
+#
+# =============================================================================
+#
+#                                   Samplers
+#
+# =============================================================================
+#
+
+class UltranestSampler(BaseSampler):
+    """This class is used to construct an Dynesty sampler from the dynesty
+    package.
+
+    Parameters
+    ----------
+    model : model
+        A model from ``pycbc.inference.models``.
+    """
+    name = "ultranest"
+    _io = UltranestFile
+
+    def __init__(self, model, **kwargs):
+        super(UltranestSampler, self).__init__(model)
+
+        import ultranest
+        log_likelihood_call, prior_call = setup_calls(model, copy_prior=True)
+
+        self._sampler = ultranest.ReactiveNestedSampler(
+            list(self.model.variable_params),
+            log_likelihood_call,
+            prior_call)
+
+        self.nlive = 0
+        self.ndim = len(self.model.variable_params)
+        self.result = None
+        self.kwargs = kwargs  # Keywords for the run method of ultranest
+
+    def run(self):
+        self.result = self._sampler.run(**self.kwargs)
+
+    @property
+    def io(self):
+        return self._io
+
+    @property
+    def niterations(self):
+        return self.result['niter']
+
+    @classmethod
+    def from_config(cls, cp, model, **kwds):
+        """
+        Loads the sampler from the given config file.
+        """
+        skeys = {}
+        opts = {'update_interval_iter_fraction': float,
+                'update_interval_ncall': int,
+                'log_interval': int,
+                'show_status': bool,
+                'dlogz': float,
+                'dKL': float,
+                'frac_remain': float,
+                'Lepsilon': float,
+                'min_ess': int,
+                'max_iters': int,
+                'max_ncalls': int,
+                'max_num_improvement_loops': int,
+                'min_num_live_points': int,
+                'cluster_num_live_points:': int}
+        for opt_name in opts:
+            if cp.has_option('sampler', opt_name):
+                value = cp.get('sampler', opt_name)
+                skeys[opt_name] = opts[opt_name](value)
+        return cls(model, **skeys)
```

Comment:
You forgot to call `setup_output`! I think you may have started this before Sumit's patch was merged. The `from_config` should take an output file keyword argument. After initializing the cls, you then run `setup_output` using the output file. See the Dynesty sampler for an example.